src/ralph/networks/models/networks.py

```diff
@@ -0,0 +1,416 @@
+            return 0
+        if self.netmask == 31:
+            return 2
+        return self.max_ip - self.min_ip - 1
+
+    class Meta:
+        verbose_name = _('network')
+        verbose_name_plural = _('networks')
+        unique_together = ('min_ip', 'max_ip')
+
+    def __str__(self):
+        return '{} ({})'.format(self.name, self.address)
+
+    def save(self, *args, **kwargs):
+        """
+        Override standard save method. Method saves min_ip and max_ip
+        depenfing on the self.address.
+
+        Example:
+            >>> network = Network.objects.create(
+            ...     name='C', address='192.168.1.0/24'
+            ... )
+            >>> network.min_ip, network.max_ip, network.gateway
+            (3232235776, 3232236031, None)
+        """
+        self.min_ip = int(self.network_address)
+        self.max_ip = int(self.broadcast_address)
+        self._assign_parent()
+        super(Network, self).save(*args, **kwargs)
+        self._assign_ips_to_network()
+
+    def _assign_ips_to_network(self):
+        IPAddress.objects.exclude(
+            network=self
+        ).filter(
+            number__gte=self.min_ip,
+            number__lte=self.max_ip
+        ).update(
+            network=self
+        )
+
+    def get_subnetworks(self):
+        return self.get_descendants()
+
+    def reserve_margin_addresses(self, bottom_count=None, top_count=None):
+        ips = []
+        exists_ips = set(IPAddress.objects.filter(
+            Q(number__gte=self.min_ip, number__lte=self.min_ip + bottom_count) |
+            Q(number__gte=self.max_ip - top_count, number__lte=self.max_ip)
+        ).values_list('number', flat=True))
+        to_create = set(chain.from_iterable([
+            range(int(self.min_ip + 1), int(self.min_ip + bottom_count)),
+            range(int(self.max_ip - top_count), int(self.max_ip - 1))
+        ]))
+        to_create = to_create - exists_ips
+        for ip_as_int in to_create:
+            ips.append(IPAddress(
+                number=ip_as_int,
+                network=self,
+                status=IPAddressStatus.reserved
+            ))
+        IPAddress.objects.bulk_create(ips)
+        return len(to_create), exists_ips - to_create
+
+    def get_first_free_ip(self):
+        used_ips = list(
+            self.ips.values_list('number', flat=True).order_by('number')
+        )
+        min_ip = int(self.min_ip if self.netmask == 31 else self.min_ip + 1)
+        max_ip = int(self.max_ip + 1 if self.netmask == 31 else self.max_ip)
+        ip_as_int = None
+        for ip_as_int in range(min_ip, max_ip):
+            if ip_as_int not in used_ips:
+                break
+        # TODO: do it better
+        last = ip_as_int in used_ips
+        return (
+            ipaddress.ip_address(ip_as_int)
+            if ip_as_int and not last else None
+        )
+
+    def search_networks(self):
+        """
+        Search networks (ancestors) order by smallest.
```

Comment:
order by smallest -> order first by min_ip descending, then by max_ip ascending, to get smallest ancestor netowork containing current network